Automattic/wp-calypso
# composite-checkout: Add full credits payment method

#### Changes proposed in this Pull Request

This adds a new payment method that allows paying for an entire purchase with credits. Partial credits payments will be handled separately. This payment method will only be included if the credits total is greater than or equal to the order total.

In order to implement this, some refactoring had to be done, which makes this PR rather unfortunately large. To make this PR easier to review, I've moved all that refactoring into its own PR #38759. 

#### New credits endpoint data

Because we're going to need to display the credit amount and also compare it to the total, the current major unit of credits we get from the shopping cart is not sufficient and I added a display and integer version in D37452-code.

#### Screenshots

<img width="570" alt="Screen Shot 2020-01-09 at 2 12 14 PM" src="https://user-images.githubusercontent.com/2036909/72097931-74699000-32eb-11ea-85f1-ad6b79c75bf8.png">

#### Testing instructions

- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`.
- Sandbox the store.
- Make sure your test account has no credits (or fewer credits than a plan cost).
- Add a plan to your cart.
- Visit checkout and verify that the "Credits" method does not appear in the first step.
- Add enough credits to your test account to cover a plan.
- Visit checkout again and verify that the "Credits" method appears, is the first payment method, and that it is selected by default.
- Complete checkout with the "Credits" payment method and verify that the payment is successful. 

Fixes #24448
